src/compute-client/src/as_of_selection.rs

```diff
@@ -0,0 +1,521 @@
+    /// inputs.
+    ///
+    /// Failing to apply this constraint to a collection is an error. The affected dataflow will
+    /// not be able to hydrate successfully.
+    fn apply_upstream_storage_constraints(
+        &self,
+        storage_read_holds: &BTreeMap<GlobalId, ReadHold<T>>,
+    ) {
+        // Apply direct constraints from storage inputs.
+        for (id, collection) in &self.collections {
+            for input_id in &collection.storage_inputs {
+                let read_hold = &storage_read_holds[input_id];
+                let constraint = Constraint {
+                    type_: ConstraintType::Hard,
+                    bound_type: BoundType::Lower,
+                    frontier: read_hold.since(),
+                    reason: &format!("storage input {input_id} read frontier"),
+                };
+                self.apply_constraint(*id, constraint);
+            }
+        }
+
+        // Propagate constraints downstream, restoring `AsOfBounds` invariant (1).
+        self.propagate_bounds_downstream(BoundType::Lower, ConstraintType::Hard);
+    }
+
+    /// Apply as-of constraints imposed by the frontiers of downstream storage collections.
+    ///
+    /// A collection's as-of _must_ be <= the join of the read frontier and the write frontier of
+    /// the storage collection it exports to, if any.
+    ///
+    /// Failing to apply this constraint to a collection is an error. The storage collection it
+    /// exports to will have times visible to readers skipped in its output, violating correctness.
+    fn apply_downstream_storage_constraints(&self) {
+        // Apply direct constraints from storage exports.
+        for id in self.collections.keys() {
+            let Ok(frontiers) = self.storage_collections.collection_frontiers(*id) else {
+                continue;
+            };
+            let upper = frontiers.read_capabilities.join(&frontiers.write_frontier);
+            let constraint = Constraint {
+                type_: ConstraintType::Hard,
+                bound_type: BoundType::Upper,
+                frontier: &upper,
+                reason: &format!("storage export {id} write frontier"),
+            };
+            self.apply_constraint(*id, constraint);
+        }
+
+        // Propagate constraints upstream, restoring `AsOfBounds` invariant (2).
+        self.propagate_bounds_upstream(BoundType::Upper, ConstraintType::Hard);
+    }
+
+    /// Propagate as-of bounds through the dependency graph, in downstream direction.
+    ///
+    /// Calling this method restores `AsOfBounds` invariant (2).
+    fn propagate_bounds_downstream(&self, bound_type: BoundType, constraint_type: ConstraintType) {
+        // We don't want to rely on a correspondence between `GlobalId` order and dependency order,
+        // so we use a fixpoint loop here.
+        fixpoint(|changed| {
+            self.propagate_bounds_downstream_inner(bound_type, constraint_type, changed);
```

Comment:
In the worst case, the run time of these fixpoint loops will be kinda quadratic with the number of dataflows. More specifically:
- If there is a long chain of dataflows on a cluster, where each dataflow depends on the previous one, then the fixpoint loop will do as many steps as the length of this dataflow.
- Each step of the fixpoint loop will so as many steps as the number of dataflows across all clusters.

I guess this is ok for now, because this will run in an acceptable time even with 1000 dataflows, which is much more than any of our current users. But in the long run, we might want to improve this in some way. For example, `changed` could be made not just a bool, but it could be a set of dataflows whose bounds have changed in the last step, and then the outer loop of `propagate_bounds_downstream_inner` could go through not all collections, but just those that have an input that changed in the last step.

Reply:
While we don't assume `GlobalId`s to be in dependency order (I think currently they always are) we can assume them to be mostly in dependency order. So I would expect each fixpoint loop only to be stepped twice (once for the changes, once to notice there are no more changes). An exception is when propagating bounds breaks one of the invariants and we need to propagate in the other direction to fix that. In this case we can iterator multiple times, depending on the shape of dataflow dependencies.

So yes, strictly speaking the fixpoint loops are quadratic (or maybe even worse with the nested fixpoint loops?). I'd hold off with optimizing this until we know it is a performance issue, since it will make the code more complicated. But seems like a good idea to ensure that coordinator bootstrapping measures the time of as-of bootstrapping (it doesn't in the current version).

Reply:
I added logging for the as-of selection runtime in the coordinator's `bootstrap` code.